packages/story-editor/src/components/panels/document/taxonomies/FlatTermSelector.js

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import { useCallback, useDebouncedCallback } from '@web-stories-wp/react';
+
+/**
+ * Internal dependencies
+ */
+import Tags from '../../../form/tags';
+import cleanForSlug from '../../../../utils/cleanForSlug';
+import { useTaxonomy } from '../../../../app/taxonomy';
+import { ContentHeading } from './shared';
+
+function FlatTermSelector({ taxonomy }) {
+  const {
+    createTerm,
+    termCache,
+    addSearchResultsToCache,
+    setSelectedTaxonomySlugs,
+    selectedSlugs,
+  } = useTaxonomy(
+    ({
+      state: { termCache, selectedSlugs },
+      actions: {
+        createTerm,
+        addSearchResultsToCache,
+        setSelectedTaxonomySlugs,
+      },
+    }) => ({
+      termCache,
+      createTerm,
+      addSearchResultsToCache,
+      setSelectedTaxonomySlugs,
+      selectedSlugs,
+    })
+  );
+
+  const handleFreeformTermsChange = useCallback(
```

Comment:
Awesome. Yea I was starting to think this should all be wrapped up in a component instead of those memoized curried callbacks. Thanks for cleaning up!